### PR TITLE
Adds deserialization for compilers and validates tests for Primitive; fixes numerical deserialization

### PR DESCRIPTION
**Pull request by JGSweets**
~~WIP until #884 and #885 merge~~

This PR does the following:
  * Adds load_from_dict for BaseCompiler (hence all)
  * Adds a test to validate deserial for Primitive (others come in subsequent tests)
  * Fixes numerical deserial bug and adds tests to validates 
  * fixes circular import issues with the json_decoder

### CODE REVIEW

File: dataprofiler/profilers/__init__.py [hunk 1]
```diff
@@ -10,3 +12,19 @@
 from .profile_builder import Profiler, StructuredProfiler, UnstructuredProfiler
 from .text_column_profile import TextColumn
 from .unstructured_labeler_profile import UnstructuredLabelerProfile
+
+# set here to avoid circular imports
+json_decoder._profiles = {
```
Comment (JGSweets) [inline dataprofiler/profilers/__init__.py:17]: circular import resolution

File: dataprofiler/profilers/categorical_column_profile.py [hunk 2]
```diff
@@ -6,7 +6,8 @@
 
 from pandas import DataFrame, Series
 
-from . import BaseColumnProfiler, utils
```
Comment (JGSweets) [inline dataprofiler/profilers/categorical_column_profile.py:9]: circular import resolution

File: dataprofiler/profilers/column_profile_compilers.py [hunk 3]
```diff
@@ -45,6 +46,7 @@ def __init__(
         if self._option_class is None:
             raise NotImplementedError("Must set the expected OptionClass.")
 
+        self.name = None
```
Comment (JGSweets) [inline dataprofiler/profilers/column_profile_compilers.py:49]: will go away with #884 merge

File: dataprofiler/profilers/column_profile_compilers.py [hunk 4]
```diff
@@ -216,6 +218,34 @@ def update_profile(
             self._profiles[profile_type].update(df_series)
         return self
 
+    @classmethod
```
Comment (JGSweets) [inline dataprofiler/profilers/column_profile_compilers.py:221]: new in this PR

File: dataprofiler/profilers/column_profile_compilers.py [hunk 5]
```diff
@@ -216,6 +218,34 @@ def update_profile(
...
+                for col_type, profile_as_dict in value.items():
+                    value[col_type] = load_column_profile(profile_as_dict)
+                # since needs to be in the same order, use _profilers to enforce
+                value = OrderedDict(
+                    {
+                        k.type: value[k.type]
+                        for k in compiler._profilers
+                        if k.type in value
+                    }
+                )
```
Comment (JGSweets) [inline dataprofiler/profilers/column_profile_compilers.py:245]: compilers necessitate order, hence this

File: dataprofiler/profilers/data_labeler_column_profile.py [hunk 6]
```diff
@@ -9,7 +9,8 @@
 
 from ..labelers.base_data_labeler import BaseDataLabeler
 from ..labelers.data_labelers import DataLabeler
-from . import BaseColumnProfiler, utils
```
Comment (JGSweets) [inline dataprofiler/profilers/data_labeler_column_profile.py:12]: circular import resolution

File: dataprofiler/profilers/graph_profiler.py [hunk 7]
```diff
@@ -12,7 +12,8 @@
 import scipy.stats as st
 
 from ..data_readers.graph_data import GraphData
-from . import BaseColumnProfiler, utils
```
Comment (JGSweets) [inline dataprofiler/profilers/graph_profiler.py:15]: circular import resolution

File: dataprofiler/profilers/json_decoder.py [hunk 8]
```diff
@@ -1,17 +1,17 @@
 """Contains methods to decode components of a Profiler."""
-from typing import Dict, Optional, Type
+from typing import TYPE_CHECKING, Dict, Optional, Type
```
Comment (JGSweets) [inline dataprofiler/profilers/json_decoder.py:2]: circular import resolution

File: dataprofiler/profilers/json_decoder.py [hunk 9]
```diff
@@ -23,23 +23,31 @@ def get_column_profiler_class(class_name: str) -> Type[BaseColumnProfiler]:
     :type class_name: str representing name of class
     :return: subclass of BaseColumnProfiler object
     """
-    profiles: Dict[str, Type[BaseColumnProfiler]] = {
```
Comment (JGSweets) [inline dataprofiler/profilers/json_decoder.py:26]: circular import resolution

File: dataprofiler/profilers/json_decoder.py [hunk 10]
```diff
@@ -23,23 +23,31 @@ def get_column_profiler_class(class_name: str) -> Type[BaseColumnProfiler]:
...
-
-    profile_class: Optional[Type[BaseColumnProfiler]] = profiles.get(class_name)
+    profile_class: Optional[Type["BaseColumnProfiler"]] = _profiles.get(class_name)
     if profile_class is None:
         raise ValueError(f"Invalid profiler class {class_name} " f"failed to load.")
     return profile_class
 
 
-def load_column_profile(serialized_json: dict) -> BaseColumnProfiler:
+def get_compiler_class(class_name: str) -> Type["col_pro_compilers.BaseCompiler"]:
```
Comment (JGSweets) [inline dataprofiler/profilers/json_decoder.py:32]: new

File: dataprofiler/profilers/json_decoder.py [hunk 11]
```diff
@@ -54,14 +62,42 @@ def load_column_profile(serialized_json: dict) -> BaseColumnProfiler:
...
-        serialized using the custom encoder in profilers.json_encoder
+        # serialized using the custom encoder in profilers.json_encoder
     :type serialized_json: a dict that was created by calling json.loads on
         a JSON representation using the custom encoder
     :return: subclass of BaseColumnProfiler that has been deserialized from
         JSON
 
     """
-    column_profiler_cls: Type[BaseColumnProfiler] = get_column_profiler_class(
+    column_profiler_cls: Type["BaseColumnProfiler"] = get_column_profiler_class(
```
Comment (JGSweets) [inline dataprofiler/profilers/json_decoder.py:72]: new

File: dataprofiler/profilers/json_encoder.py [hunk 12]
```diff
@@ -24,6 +24,7 @@ def default(self, to_serialize):
             (
                 base_column_profilers.BaseColumnProfiler,
                 numerical_column_stats.NumericStatsMixin,
+                column_profile_compilers.BaseCompiler,
```
Comment (JGSweets) [inline dataprofiler/profilers/json_encoder.py:27]: will go away with #884 merge

File: dataprofiler/profilers/numerical_column_stats.py [hunk 13]
```diff
@@ -403,7 +403,7 @@ def convert_histogram_key_types_to_np(histogram_info: dict):
         if self.num_zeros is not None:
             self.num_zeros = np.int64(self.num_zeros)
         if self.num_negatives is not None:
-            self.num_negatives = np.int64(self.num_zeros)
+            self.num_negatives = np.int64(self.num_negatives)
```
Comment (JGSweets) [inline dataprofiler/profilers/numerical_column_stats.py:406]: fix to numerical deserial

File: dataprofiler/profilers/order_column_profile.py [hunk 14]
```diff
@@ -5,7 +5,8 @@
 
 from pandas import DataFrame, Series
 
-from . import BaseColumnProfiler, utils
```
Comment (JGSweets) [inline dataprofiler/profilers/order_column_profile.py:8]: circular import resolution

File: dataprofiler/profilers/unstructured_text_profile.py [hunk 15]
```diff
@@ -9,7 +9,8 @@
 from numpy import ndarray
 from pandas import DataFrame, Series
 
-from . import BaseColumnProfiler, utils
```
Comment (JGSweets) [inline dataprofiler/profilers/unstructured_text_profile.py:12]: circular import resolution

File: dataprofiler/profilers/utils.py [hunk 16]
```diff
@@ -14,6 +14,7 @@
 from itertools import islice
 from multiprocessing.pool import Pool
 from typing import (
+    TYPE_CHECKING,
```
Comment (JGSweets) [inline dataprofiler/profilers/utils.py:17]: circular import resolution

File: dataprofiler/tests/profilers/test_base_column_profilers.py [hunk 17]
```diff
@@ -163,7 +163,7 @@ def test_json_encode(self):
             }
         )
 
-        self.assertEqual(serialized, expected)
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_base_column_profilers.py:166]: will go away with #884 or #885

File: dataprofiler/tests/profilers/test_column_profile_compilers.py [hunk 18]
```diff
@@ -81,6 +87,179 @@ def test_add_profilers(self):
         self.assertEqual(3, merged_compiler._profiles["test"])
         self.assertEqual("compiler1", merged_compiler.name)
 
+    def test_compiler_stats_diff(self):
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_column_profile_compilers.py:90]: will go away with #885

File: dataprofiler/tests/profilers/test_column_profile_compilers.py [hunk 19]
```diff
@@ -305,144 +484,97 @@ def test_disabling_columns_during_primitive_diff(self):
         # Test disabling all columns in all compilers
         compiler2 = col_pro_compilers.ColumnPrimitiveTypeProfileCompiler(data2, options)
         expected_diff = {}
-        self.assertDictEqual(expected_diff, compiler1.diff(compiler2))
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_column_profile_compilers.py:308]: will go away with #885

File: dataprofiler/tests/profilers/test_column_profile_compilers.py [hunk 20]
```diff
@@ -305,144 +484,97 @@ def test_disabling_columns_during_primitive_diff(self):
...
+                },
             }
-            compiler1._profiles["data_labeler"].label_representation = {
-                "a": 0.15,
-                "b": 0.01,
-                "c": 0.84,
+        )
+        self.assertEqual(expected, serialized)
+
+    def test_json_encode_after_update(self):
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_column_profile_compilers.py:505]: # will go away with #884

File: dataprofiler/tests/profilers/test_column_profile_compilers.py [hunk 21]
```diff
@@ -305,144 +484,97 @@ def test_disabling_columns_during_primitive_diff(self):
...
-        expected_diff = {}
-        self.assertDictEqual(expected_diff, compiler1.diff(compiler2))
+        test_utils.assert_profiles_equal(expected_compiler, deserialized)
 
-    @mock.patch.multiple(col_pro_compilers.BaseCompiler, __abstractmethods__=set())
-    def test_no_profilers_error(self):
-        with self.assertRaises(NotImplementedError) as e:
-            col_pro_compilers.BaseCompiler()
-        self.assertEqual("Must add profilers.", str(e.exception))
+    def test_json_decode_after_update(self):
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_column_profile_compilers.py:545]: new, why weird git.. idk

File: dataprofiler/tests/profilers/test_column_profile_compilers.py [hunk 22]
```diff
@@ -305,144 +484,97 @@ def test_disabling_columns_during_primitive_diff(self):
...
 
-            expected_diff = {
-                "statistics": {
-                    "avg_predictions": {"a": "unchanged", "b": -0.7, "c": 0.7},
-                    "label_representation": {"a": -0.84, "b": "unchanged", "c": 0.84},
-                },
-                "data_label": [["a"], [], ["b"]],
-            }
-            self.assertDictEqual(expected_diff, compiler1.diff(compiler2))
+    def test_json_decode(self):
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_column_profile_compilers.py:537]: new, git being weird

File: dataprofiler/tests/profilers/test_float_column_profile.py [hunk 23]
```diff
@@ -1830,7 +1830,7 @@ def test_json_encode(self):
                 },
             }
         )
-        self.assertEqual(serialized, expected)
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_float_column_profile.py:1833]: will go away with #885

File: dataprofiler/tests/profilers/test_float_column_profile.py [hunk 24]
```diff
@@ -1956,7 +1956,7 @@ def test_json_encode_after_update(self, time):
             }
         )
 
-        self.assertEqual(serialized, expected)
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_float_column_profile.py:1959]: will go away with #885

File: dataprofiler/tests/profilers/test_float_column_profile.py [hunk 25]
```diff
@@ -1972,9 +1972,7 @@ def test_json_decode_after_update(self):
         # Actual deserialization
 
         # Build expected FloatColumn
-        df_float = pd.Series([1.5, 2.2, 5.0, 7.0, 4.0, 3.0, 2.0, 7.0, 8.0, 9.0]).apply(
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_float_column_profile.py:1975]: fix to ensure test cov

File: dataprofiler/tests/profilers/test_int_column_profile.py [hunk 26]
```diff
@@ -1309,7 +1309,7 @@ def test_json_decode_after_update(self):
         # Actual deserialization
 
         # Build expected IntColumn
-        df_int = pd.Series([1, 2, 5, 7, 4, 3, 2, 7, 8, 9])
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_int_column_profile.py:1312]: fix to ensure test cov

File: dataprofiler/tests/profilers/utils.py [hunk 27]
```diff
@@ -183,8 +183,9 @@ def assert_profiles_equal(actual, expected):
     actual_dict = actual.__dict__ if not isinstance(actual, dict) else actual
     expected_dict = expected.__dict__ if not isinstance(expected, dict) else expected
 
-    assert len(actual_dict.keys()) == len(expected_dict.keys())
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/utils.py:186]: len is not needed as key check does len by virtual of matching keys

File: dataprofiler/tests/profilers/utils.py [hunk 28]
```diff
@@ -201,7 +202,7 @@ def assert_profiles_equal(actual, expected):
         elif isinstance(actual_value, dict):
             assert_profiles_equal(actual_value, expected_value)
         elif isinstance(actual_value, numbers.Number):
-            np.testing.assert_equal(actual_value, expected_value)
+            np.testing.assert_equal(actual_value, expected_value, f"{key}")
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/utils.py:205]: clarity on failure, how i found the `num_negative issue`.

Comment (tyfarnan) [reply dataprofiler/tests/profilers/utils.py:205]: nice

**Review by tyfarnan — APPROVED** [review]

**Review by ksneab7 — APPROVED** [review]

**Review by taylorfturner — APPROVED** [review]

**Pull request merged**